src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,202 @@
+    }
+
+    /// Budget is increased for each key atomically in Redis
+    fn increase_budget_consumed(&mut self, packets: &[Ciphertext], budget: i32) {
+        for packet in packets.iter() {
+            let _: () = self
+                .redis_connection
+                .incr(packet.to_string(), budget)
+                .expect("failed to execute INCR for a packet");
+        }
+    }
+
+    fn get_eligibility_of_packets(
+        &mut self,
+        packets: &[Ciphertext],
+    ) -> (Vec<Ciphertext>, Vec<Ciphertext>) {
+        let (mut eligible_packets, mut ineligible_packets): (Vec<Ciphertext>, Vec<Ciphertext>) =
+            (Vec::new(), Vec::new());
+        for packet in packets.iter() {
+            let budget_consumed: i32 = self
+                .redis_connection
+                .get(packet.to_string())
+                .expect("Failed to GET budget for packet");
+            if budget_consumed <= self.total_privacy_budget {
+                eligible_packets.push(packet.clone());
+            } else {
+                ineligible_packets.push(packet.clone());
+            }
+        }
+        (eligible_packets, ineligible_packets)
+    }
+
+    /// Budget is decreased for each ineligible key atomically in Redis
+    fn decrease_budget_consumed_for_ineligible_packets(
+        &mut self,
+        packets: Vec<Ciphertext>,
+        budget: i32,
+    ) {
+        for packet in packets.iter() {
+            let _: () = self
+                .redis_connection
+                .decr(packet.to_string(), budget)
+                .expect("failed to execute DECR for a packet");
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::thread_rng;
+    use rust_elgamal::{Ciphertext, DecryptionKey, RistrettoPoint, Scalar};
+    use std::env;
+
+    fn create_ciphertexts() -> [Ciphertext; 4] {
+        let mut rng = thread_rng();
+
+        let k1 = DecryptionKey::new(&mut rng);
+        let mk1 = RistrettoPoint::random(&mut rng);
+        let mk2 = RistrettoPoint::random(&mut rng);
+        let mk3 = RistrettoPoint::random(&mut rng);
+        let mk4 = RistrettoPoint::random(&mut rng);
+        let r = Scalar::random(&mut rng);
+
+        let c1 = k1.encryption_key().encrypt_with(mk1, r);
+        let c2 = k1.encryption_key().encrypt_with(mk2, r);
+        let c3 = k1.encryption_key().encrypt_with(mk3, r);
+        let c4 = k1.encryption_key().encrypt_with(mk4, r);
+
+        [c1, c2, c3, c4]
+    }
+
+    fn clear_redis_keys(privacy_budget: &mut PrivacyBudget, packets: &[Ciphertext]) {
+        for packet in packets.iter() {
+            let _: () = privacy_budget
+                .redis_connection
+                .del(packet.to_string())
+                .expect("Failed to delete keys");
+        }
+    }
+
+    fn create_privacy_budget_ciphertexts() -> (PrivacyBudget, [Ciphertext; 4]) {
+        let redis_host_name =
+            env::var("REDIS_HOSTNAME").expect("missing environment variable REDIS_HOSTNAME");
```

Comment:
This is where things get tedious.  It's not ideal if someone who runs the tool has to start their own redis server, get it into the expected state, and set environment variables.

What these tests need to do is start the redis server on their own.  I know that's a massive burden, because it involves a bunch of tricky setup (I suggest using the lazy_static crate for executing the database).  The `redis` crate has [a good example of how you might set that up](https://github.com/mitsuhiko/redis-rs/blob/master/tests/support/mod.rs) (you don't need all that complexity, so that can be pared back fairly aggressively).

Then your only requirement for running the tests is that you have the redis server installed.

Reply:
Have implemented a setup and teardown  service for each test which starts and stops the redis server. However, while running tests, the tests keep passing and failing randomly in different runs. 
Also tried to run all the tests in a single thread but that didnt help alleviate the issue.
Let me know your thoughts